### PR TITLE
P0919R3 Heterogeneous lookup for unordered containers

### LINKED ISSUE #2433: [2018-11 LWG Motion 27] P0919R3 Heterogeneous lookup for unordered containers
[P0919R3](http://wg21.link/p0919r3)

### PR DESCRIPTION
**Pull request by burblebee**
Fixes #2433.

### CODE REVIEW

File: source/containers.tex [hunk 1]
```diff
@@ -2268,6 +2272,12 @@
 \tcode{k} denotes a value of type \tcode{key_type},
 \tcode{hf} denotes a possibly const value of type \tcode{hasher},
 \tcode{eq} denotes a possibly const value of type \tcode{key_equal},
+\tcode{ke} is a value such that
+(1) \tcode{eq(r1, ke) == eq(ke, r1)}
+with \tcode{r1} the key value of \tcode{e} and \tcode{e} in \tcode{a_tran},
```
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex]: PS: not sure what is "outdated" about this comment.  I did type it in a while ago without realizing I needed to "submit" it, but the branch hasn't changed, so it still applies.

**Comment by cpplearner** [comment]
The following uses of `Pred` should probably be changed to use `key_equal` (or "key equality predicate").

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L2168

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L2351-L2353

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L2939

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L2944

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L3023

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L3034

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L7619

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L7677

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L8177

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L8218

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L8518

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L8557

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L8819

https://github.com/cplusplus/draft/blob/c1140dc035aef43baeb9a4f1daea3dfbc6201c5b/source/containers.tex#L8858

**Comment by zygoloid** [comment]
@cpplearner I think you might be right, but it's not clear to me that that's an editorial change. In particular, the behavior of unordered containers with a hasher that has a `transparent_key_equal` and whose `Pred` is `equal_to<Key>` (which is the only possibility where it can be different from the `transparent_key_equal` type) seems highly underspecified to me. (Is the specified `Pred` entirely ignored, or sometimes still used? What happens if `operator==` is inconsistent with the "transparent" key equality functor? How does eg `unordered_set<std::string, hash_by_lowercased_contents>` behave?)

Can you ask to get a library issue opened to replace `Pred` with `key_equal` in all the relevant places?

File: source/containers.tex (hunk 1 above)
[the remark this answers is not in the document]
Comment (burblebee) [reply source/containers.tex]: @zygoloid fixed the above wording in a followup commit.

**Pull request merged**